File: pkg/models/game.go
package models

import (
	"fmt"
)

// DominoTile represents a single domino piece.
type DominoTile struct {
	UID        string `json:"uid,omitempty"`
	ID         string `json:"tile_id,omitempty"`
	ValueLeft  int    `json:"value_left"`
	ValueRight int    `json:"value_right"`
}

// NewTile is a constructor that auto-generates the deterministic ID.
func NewTile(left, right int) DominoTile {
	return DominoTile{
		ID:         fmt.Sprintf("%d-%d", left, right),
		ValueLeft:  left,
		ValueRight: right,
	}
}

// IsDouble returns true if both sides of the domino have the same value.
func (t DominoTile) IsDouble() bool {
	return t.ValueLeft == t.ValueRight
}

// String provides a clean string representation for debugging or logging.
func (t DominoTile) String() string {
	return fmt.Sprintf("[%d|%d]", t.ValueLeft, t.ValueRight)
}

// PlayerHand represents the private state of a player's tiles during a game.
type PlayerHand struct {
	UID       string       `json:"uid,omitempty"`
	PlayerID  string       `json:"player_id"`
	Tiles     []DominoTile `json:"tiles"`
	HasPassed bool         `json:"has_passed"`
	IsReady   bool         `json:"is_ready"`
}

// GameSession represents the complete state of an active match.
type GameSession struct {
	UID            string       `json:"uid,omitempty"`
	SessionID      string       `json:"session_id,omitempty"`
	Status         string       `json:"status"`
	Players        []string     `json:"players"`
	Hands          []PlayerHand `json:"hands"`
	Boneyard       []DominoTile `json:"boneyard"`
	GameBoard      []DominoTile `json:"game_board"`
	LeftOpenValue  int          `json:"left_open_value"`
	RightOpenValue int          `json:"right_open_value"`
	CurrentTurn    string       `json:"current_turn"`

	// Dgraph-specific persistence blobs
	BoneyardRaw  string `json:"boneyard_raw,omitempty"`
	GameBoardRaw string `json:"game_board_raw,omitempty"`
}

File: web/src/types/game.ts
export interface DominoTile {
  tile_id: string
  value_left: number
  value_right: number
  uid?: string
}

export interface PlayerHand {
  player_id: string
  tiles?: DominoTile[]
  tile_count?: number
  has_passed: boolean
  is_ready: boolean
  is_abandoned: boolean
}

export interface GameSession {
  session_id?: string
  status: string
  players: string[]
  hands: PlayerHand[]
  boneyard_count?: number
  game_board: DominoTile[]
  left_open_value: number
  right_open_value: number
  current_turn: string
  mutations_locked?: boolean
}

export interface MatchStateResponse {
  match_id: string
  player_id: string
  found: boolean
  session?: GameSession
  node_id?: string
  state?: string
  error?: string
}

export interface MatchActionData {
  kind: 'PLAY_TILE' | 'DRAW' | 'PASS'
  player_id: string
  tile?: DominoTile
  play_at_left?: boolean
}

export interface MatchMutateRequest {
  client_id: string
  sequence_number: number
  match_id: string
  player_id: string
  action_data: MatchActionData
}

export interface MatchMutateResponse {
  ok: boolean
  idempotent_replay?: boolean
  result?: {
    ok: boolean
    session?: GameSession
    applied?: boolean
  }
  error?: string
}

export interface EventEnvelope<T = unknown> {
  type: string
  timestamp: number
  payload: T
}

export interface SessionDeltaPayload {
  session_id: string
  match_id: string
  op: string
  applied?: boolean
  session?: GameSession
}

export interface PlayOption {
  left: boolean
  right: boolean
}
